### ISSUE TITLE
`nixos-rebuild switch --upgrade` over ssh may put system into an unusable state

### ISSUE THREAD

**Issue by nagisa** [NixOS/nixpkgs#39118]
## Issue description

Upgrading with `nixos-rebuild switch --upgrade` may put system into an unusable state by disabling services and terminating the SSH session, thus killing the `rebuild` process(?)

```
$ sudo nixos-rebuild switch --upgrade
unpacking channels...
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/ngm996cnslb1yz15g09hcd8bzhsln138-system-path.drv
  /nix/store/0b3agdfcv1ax3qn9a2bi4ri4c5a0wrpm-dbus-1.drv
  /nix/store/bcy7asyvxa5xza1bdy5ph8qx0viadfmr-unit-dbus.service.drv
  /nix/store/8v6mdb50nbr73x3k18v39cg2wyiahbn0-user-units.drv
  /nix/store/i3q3ycxzpnqyy811c4l67p75pbl5spvp-unit-polkit.service.drv
  /nix/store/rp6j33aazxv86pm5j6r4apc9pqknpdib-unit-systemd-fsck-.service.drv
  /nix/store/hpgr2fllnbmmjhpb74ip1wlb8zrj63y2-system-units.drv
  /nix/store/yc462a9jb3qwsc9yfygdwcyk35krbssd-etc.drv
  /nix/store/8hl7rjwzd27zasf87nxlj6jkwfcpkmpv-nixos-system-shirobox-18.09pre137207.92f0d31b949.drv
these paths will be fetched (194.31 MiB download, 1694.29 MiB unpacked):
  /nix/store/3cv3s9lv79qjkqyafbclbk03gz5z813w-ghc-8.2.2
  /nix/store/9vgm81xw5z9psmizacwbi83c88xabam8-clang-wrapper-5.0.1
  /nix/store/nz60fx1vyva1a2rsw0jp0qsrjq91xhbw-clang-5.0.1
copying path '/nix/store/3cv3s9lv79qjkqyafbclbk03gz5z813w-ghc-8.2.2' from 'https://cache.nixos.org'...
copying path '/nix/store/nz60fx1vyva1a2rsw0jp0qsrjq91xhbw-clang-5.0.1' from 'https://cache.nixos.org'...
copying path '/nix/store/9vgm81xw5z9psmizacwbi83c88xabam8-clang-wrapper-5.0.1' from 'https://cache.nixos.org'...
building '/nix/store/ngm996cnslb1yz15g09hcd8bzhsln138-system-path.drv'...
Use of uninitialized value $propagated in split at /nix/store/9g4wc31j7a2xp22xpgwr0qssfxahxdzl-builder.pl line 146.
created 6358 symlinks in user environment
building '/nix/store/0b3agdfcv1ax3qn9a2bi4ri4c5a0wrpm-dbus-1.drv'...
building '/nix/store/i3q3ycxzpnqyy811c4l67p75pbl5spvp-unit-polkit.service.drv'...
building '/nix/store/rp6j33aazxv86pm5j6r4apc9pqknpdib-unit-systemd-fsck-.service.drv'...
building '/nix/store/bcy7asyvxa5xza1bdy5ph8qx0viadfmr-unit-dbus.service.drv'...
building '/nix/store/hpgr2fllnbmmjhpb74ip1wlb8zrj63y2-system-units.drv'...
building '/nix/store/8v6mdb50nbr73x3k18v39cg2wyiahbn0-user-units.drv'...
building '/nix/store/yc462a9jb3qwsc9yfygdwcyk35krbssd-etc.drv'...
building '/nix/store/8hl7rjwzd27zasf87nxlj6jkwfcpkmpv-nixos-system-shirobox-18.09pre137207.92f0d31b949.drv'...
perl: warning: Setting locale failed.
perl: warning: Please check that your locale settings:
	LANGUAGE = (unset),
	LC_ALL = (unset),
	LANG = "en_GB.UTF-8"
    are supported and installed on your system.
perl: warning: Falling back to the standard locale ("C").
stopping the following units: audit.service, cpufreq.service, kmod-static-nodes.service, network-link-enp24s0.service, network-local-commands.service, network-setup.service, nix-daemon.service, nix-daemon.socket, nscd.service, rngd.service, rtkit-daemon.service, samba-nmbd.service, samba-smbd.service, samba-winbindd.service, system.slice, systemd-modules-load.service, systemd-sysctl.service, systemd-timesyncd.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, transmission.service, zfs-import-sb.service, zfs-scrub.timer, zfs-snapshot-daily.timer, zfs-snapshot-frequent.timer, zfs-snapshot-hourly.timer, zfs-snapshot-monthly.timer, zfs-snapshot-weekly.timer
Connection to 192.168.1.5 closed by remote host.
Connection to 192.168.1.5 closed.
```

### Steps to reproduce

No idea. I ran `nixos-rebuild switch --upgrade` over a ssh session.

## Technical details

> Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the results.

Can’t access my machine now.

**Comment by nixy** [comment]
Since it tried to restart networking this upgrade probably won't be able to be done using `nixos-rebuild switch` over an SSH connection.

The `nixos-rebuild` command can also operate in `boot` and `test` modes which should help you work around this. If you do `nixos-rebuild boot --upgrade` then it won't switch into the new configuration till you reboot (where it would have to start up networking again anyways) and `nixos-rebuild test --upgrade` would immediately switch you into it without making it the boot default.

You may need to try rolling back and using the boot option to get back to a working state.

**Comment by nagisa** [comment]
I interface with this machine chiefly over ssh and since this is a local network, I don’t have a pretext to be worried about network breaking, which is why I don’t tend to use something like nohup or tmux. This is the first time over, I think, a year of nixos use that `nixos-rebuild switch` decides to restart networking in such a crude manner – it never happened before even after I changed network configuration.

This time, however, no network configuration changes have been made so I had no reason to believe nixos-rebuild would restart networking (it seems it actually restarted *every* service there is) this time either.

It would be ideal to me if there was a way to block `nixos-rebuild switch` *if* it would involve killing networking or the ssh daemon – otherwise this will definitely hit me again when I expect this the least.

---

I’ll keep in mind to only ever use `--upgrade` in conjunction with `boot`.

**Comment by ElvishJerricco** [comment]
It might be good if nixos-rebuild did the activation in a `systemd-run`. This way all forms of TTY terminating are non-damaging. I'd still want the activation log to go to the TTY though; not sure if that's possible.

**Comment by ElvishJerricco** [comment]
@nagisa In the meantime, you can probably just do `alias nixos-rebuild="systemd-run -t --wait /run/current-system/bin/nixos-rebuild"`. I *believe* that should be resilient to this kind of failure.

**Comment by deliciouslytyped** [comment]
https://github.com/NixOS/nixpkgs/issues/34902

**Comment by Thesola10** [comment]
Still important I think

**Comment by keithy** [comment]
https://gitlab.com/angelbox/nix/nixbox/-/blob/main/action/lib/rebuild

```
#!/usr/bin/env bash

systemd-run -E NIX_PATH \
            -E LOCALE_ARCHIVE \
            -t --wait /run/current-system/sw/bin/nixos-rebuild $@
```

**Comment by thiagokokada** [comment]
Opened a PR to fix this issue: https://github.com/NixOS/nixpkgs/pull/258571

**Comment by Mastermjr** [comment]
I just ran into this today going from 23.05 -> 23.11 over ssh, 
In the future, I'm going to try: 
```
systemd-run -t --wait /run/current-system/sw/bin/nixos-rebuild"
```
in a tmux session and that way the command runs on the server, which should hopefully prevent this issue. 
I'll report back if it does not

**Comment by thiagokokada** [comment]
This issue should not happen anymore since https://github.com/NixOS/nixpkgs/pull/258571 got merged, but I don't think this is available on 23.11.

Will leave this issue open for a while just to make sure that there are no new reports, but eventually we should close this issue.

**Comment by nagisa** [comment]
Coincidentally while I have gotten into a habbit of using `boot + kexec` for activation I also haven't observed things going awry if I accidentally `switch` so I think its fine to close, yeah.

**Comment by thiagokokada** [comment]
> Coincidentally while I have gotten into a habbit of using `boot + kexec` for activation I also haven't observed things going awry if I accidentally `switch` so I think its fine to close, yeah.

Ok, I will close the issue, but if someone is still having it in unstable channel please ping me and I can reopen it.

**Comment by ElvishJerricco** [comment]
> This issue should not happen anymore since https://github.com/NixOS/nixpkgs/pull/258571 got merged, but I don't think this is available on 23.11.

For the record, it *is* on 23.11

**Comment by ElvishJerricco** [comment]
@9999years Can you explain that more? The PR that fixes this was merged and is in 23.11. In what way is it not working?

**Comment by SuperSandro2000** [comment]
23.11 is EOL and that is definitely on 24.05, so I think we can close this.
